Assets/MixedRealityToolkit/Services/BaseDataProviderAccessCoreSystem.cs

```diff
@@ -19,47 +19,89 @@ public abstract class BaseDataProviderAccessCoreSystem : BaseCoreSystem, IMixedR
     {
         private readonly List<IMixedRealityDataProvider> dataProviders = new List<IMixedRealityDataProvider>();
 
+        /// <inheritdoc />
         public override void Reset()
         {
             base.Reset();
 
-            foreach(var provider in dataProviders)
+            try
             {
-                provider.Reset();
+                foreach (var provider in dataProviders)
```

Comment:
The try/catch here seems like it'll prevent exceptions for leaking out, but I'm not sure what this means for the rest of the data providers that are beyond the exception - i.e. if we're in the middle of resetting and we don't get to call it for the rest, don't we put ourselves into a potentially bad state (i.e. some data providers are never invoked?)

I think this is probably more critical for things like "Destroy" where we should be calling destroy on each of the dataproviders, so maybe I should have left this comment down below.

Could you elaborate a little more on why this is safe, and how in cases like Destroy we're ensuring that all of the data providers actually have the chance to get destroyed?

Reply:
The end-behavior would be the same. If the collection is changed during enumeration, the exception is thrown and the loop aborts. Adding the try/catch, as you mentioned, keeps the exception from bothering the customer as well as provides MRTK a way of adding future messaging, etc.

Reply:
I think what I'm trying to understand here is, if Destroy() is happening and an exception is thrown, I don't see how the other data providers' Destroy() function is ever called - I'm guessing that this change was done because we hit an exception in testing some of these things, and in the case of Update() it's probably not problematic, but I'm trying to understand how the Destroy() case specifically recovers - i.e. will Destroy() get queued up and called on the next frame when it's safe?

My assertion here is that I think that if an exception is thrown on Destroy, there is nothing in the system that will ensure that the other data providers' destroy() function will be called - if my assertion isn't correct, I just want to understand why I'm wrong.

Reply:
you have a good point here. we can add an additional exception within the loop.

regarding current behavior, i believe an exception in Destroy() would also halt the loop

Reply:
added exception handling in the loops to protect against a provider throwing and causing the loop to abort.

also limited the outer try/catch to Update and LateUpdate, with corresponding comment blocks to explain